### PR TITLE
lang: add cmakescript language

### LINKED ISSUE #5502: [REQUEST]: CMake v3.27
### Language name

CMAKE

### Language version

CMAKE 3.27

### Language homepage

https://cmake.org/

### Compiler homepage

https://cmake.org/

### Compiler version

3.27

### Motivation

CMake language is heavily used in C/C++ projects by the cmake interpreter. Adding support in godbolt to cmake can be trully useful for the community.

To process a cmake language script in cmake use `cmake -P script_path`

### PR DESCRIPTION
**Pull request by vicentebolea**
Fixes #5502 

In this PR I make an early attempt to add the CMakeScript as a language in compiler-explorer. I will be fantastic having this since CMake is one of the most used build system. Having an cmake interpreter in the web would probe truly useful for teams and individuals, it will be a fantastic way to share snippets of cmakescript and show how they are executed with different cmake versions.

### CODE REVIEW

**Comment by vicentebolea** [comment]
@partouf

File: lib/compilers/cmakescript.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
+// Copyright (c) 2021, Compiler Explorer Authors
```
Comment (mattgodbolt) [inline lib/compilers/cmakescript.ts:1]: Can you set this to 2023 please

File: static/modes/asmcmakescript-mode.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
+// Copyright (c) 2021, Compiler Explorer Authors
```
Comment (mattgodbolt) [inline static/modes/asmcmakescript-mode.ts:1]: Ditto

**Review by mattgodbolt — COMMENTED** [review]
Nice! Some minor things, and ... can you add one example too please? That way when folks select CMakeScript they'll see at least one example.

**Comment by vicentebolea** [comment]
Thanks @mattgodbolt for the quick review, I wonder if can you enable the CI to test this.

File: lib/compilers/_all.ts [hunk 3]
```diff
@@ -99,6 +99,7 @@ export {PythonCompiler} from './python.js';
 export {RacketCompiler} from './racket.js';
 export {RGACompiler} from './rga.js';
 export {RubyCompiler} from './ruby.js';
+export {CMakeScript} from './cmakescript.js';
```
Comment (partouf) [inline lib/compilers/_all.ts:102]: Should be CMakeScriptCompiler

**Comment by vicentebolea** [comment]
@partouf @mattgodbolt @DavidSpickett I have addressed all the comments here and it seems to run properly in my machine. Few questions

- It will be great to have a couple of versions deployed at godbolt so that the user can check how cmake behaves in different versions.
- `cmake -P` is a purely interpreted language meaning that we do not produce any type of compilation output. In that sense I have tweak the the CmakeCompiler to do the same thing in both compilation and execution phase (I made sure that it accepts users options in both cases).

Please let me know what you think, I am happy to keep improving this.

On another note from the CMake team they are happy with the naming of cmakescript as the language that we interpret with `cmake -P`

Thanks!

File: examples/cmakescript/default.cmake [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# CMake can execute standalone CMake scripts with the -P parameter
```
Comment (mattgodbolt) [inline examples/cmakescript/default.cmake:1]: Nice

**Comment by mattgodbolt** [comment]
> It will be great to have a couple of versions deployed at godbolt so that the user can check how cmake behaves in different versions.

That's all under your control! Our infra repro controls the versions of CMake that are installed, and by suppling a `cmakescript.amazon.properties` file you can configure what versions are visible.

Though we might need to tweak a little thing in the infra installer (cc @partouf ) as currently we assume there's exactly one.

**Comment by vicentebolea** [comment]
@mattgodbolt I see!

I wonder if you could approve the CI workflow.

**Review by mattgodbolt — COMMENTED** [review]
Looks ok. I'll test locally and then we can get this merged. It won't show up on the site until we get a `cmakescript.amazon.properties` and solve the installation issues but, we're close!

**Review by mattgodbolt — APPROVED** [review]
Look good! I'll take a look at the installation thing

**Comment by mattgodbolt** [comment]
@vicentebolea I'll configure some CMakes for the amazon setup :)

**Comment by vicentebolea** [comment]
> @vicentebolea I'll configure some CMakes for the amazon setup :)

Thanks! We have an internal meeting in the cmake project this Wednesday it will be super awesome to show this to the team :zap:

**Comment by mattgodbolt** [comment]
I'm trying to get this rolled out tonight :) I'll follow up here

**Comment by mattgodbolt** [comment]
This is now live!

**Pull request merged**